### PR TITLE
Fix #104: Rate-limit comment submissions

### LINKED ISSUE #104: Discord does not comment on all incompatibilities
When a user submits comments too quickly, the account is temporarily prevented from adding more. GitHub prints a message explaining this if you try to manually add a comment after running Discord. This restriction explains why Discord does not comment on all incompatibilities.

I was able to make Discord [comment on all incompatibilities](https://github.com/openjck/discord-test/pull/37/files) by [pausing for a random amount of time between each comment](https://github.com/openjck/discord/blob/%25reference-fix-missing-comments-poc/hook.js#L66-69) but that's just a proof-of-concept. We'll need a much different implementation for production. Some kind of queue would be great, but I can't find a node queue that can be configured to pause between tasks.

### PR DESCRIPTION
**Pull request by openjck**
Testing instructions:
1. Push this branch to your personal Heroku
2. `heroku config:set COMMENT_WAIT=50`
3. Push [this configuration](https://raw.githubusercontent.com/openjck/discord-test/master/.doiuse) to your test repo
4. Submit a pull request to add [this file](https://raw.githubusercontent.com/openjck/discord-test/master/incompatible.css)

Discord should comment on every incompatible line, with 20 comments total. It might take a few seconds for all of the comments to appear.

### CODE REVIEW

**Comment by openjck** [comment]
One weird thing is happening. If the value of `COMMENT_WAIT` is increased to say `100` or higher, some of the comments don't come through. If you set the value very high (say `10000`) none of the comments come through.

Do you know why that might be? I'm guessing it has something to do with `removeTokens` being hit multiple times before the first wait expires.

File: config.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 var config = {
     userAgent: 'Discord',
     token: process.env.OAUTH_TOKEN,
+    commentWait: process.env.COMMENT_WAIT,
```
Comment (darkwing) [inline config.js]: We should add a default here, like:

```
commentWait: process.env.COMMENT_WAIT || 200
```

Comment (openjck) [inline config.js]: Definitely. Will do.

**Comment by openjck** [comment]
Updated. [This comment](https://github.com/mdn/discord/pull/112#issuecomment-121403928) still applies though. Haven't found a solution yet.

File: config.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 var config = {
     userAgent: 'Discord',
     token: process.env.OAUTH_TOKEN,
+    commentWait: process.env.COMMENT_WAIT || 50,
```
Comment (groovecoder) [inline config.js]: Assume this is 50 ms?

Comment (openjck) [inline config.js]: Yep

**Comment by groovecoder** [comment]
I was able to verify this, and I can confirm a comment or 2 still get dropped, even with the 50ms wait. Hmm ...

**Comment by darkwing** [comment]
There's something I don't really trust about this.  Its core components haven't been updated in a few years and we're already seeing comments being dropped off with simple testing.

If I'm also seeing the code correctly, there's no database/queueing, which also concerns me.

**Comment by openjck** [comment]
> There's something I don't really trust about this. Its core components haven't been updated in a few years and we're already seeing comments being dropped off with simple testing.

Definitely. Limiter does have many users despite being old, but we can also think about using a different module. I'm not sure what's causing the comments to be dropped but have some suspicions. I would love to help investigate.

We could use RabbitMQ now, but a simpler in-memory rate limiter should work well enough for now given the number of users we have. I definitely agree that queuing is crucial to scale in the long term.

**Comment by openjck** [comment]
Just to be clear, this [isn't related to the GitHub API rate-limit](https://github.com/mdn/discord/issues/87#issuecomment-120193773). The code here works around a separate spam-prevention measure that GitHub activates when an account comments too quickly.

**Comment by darkwing** [comment]
Is this still relevant?

**Comment by openjck** [comment]
Definitely worth considering, yeah. This is the alternative to RabbitMQ.

**Comment by mrmakeit** [comment]
I think @darkwing is refering to #124, which uses setTimeout and checks for a successful comment.

**Comment by openjck** [comment]
Definitely. It looks like they both do similar work. This one uses the `limiter` module but doesn't check that the comment was successful. If we decide not to use Rabbit I'd love to collaborate and bring together the best of each of these.

**Comment by openjck** [comment]
Superseded by #131

**Pull request closed without merging**